### PR TITLE
WIP

### LINKED ISSUE #11416: Support responsive images
This is a large piece of work that probably depends on making our image loading code better reflect the actual specification model. For more context on what responsive images are, see [the specification introduction](https://html.spec.whatwg.org/multipage/embedded-content.html#introduction-3).

Components:
- [x] [`<picture>`](https://html.spec.whatwg.org/multipage/embedded-content.html#the-picture-element)
- [x] [`<source>`](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element)
- [x] [sizes attribute](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-sizes) (see #10989 for prior work)
- [x] [srcset attribute](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-srcset) (see #10829 for prior work)
- [x] [relevant mutations](https://html.spec.whatwg.org/multipage/embedded-content.html#relevant-mutations) processing
- [x] [update the source set](https://html.spec.whatwg.org/multipage/embedded-content.html#update-the-source-set) as part of the image selection process
- [x] [environment changes](https://html.spec.whatwg.org/multipage/embedded-content.html#img-environment-changes) processing

### PR DESCRIPTION
**Pull request by Rakhisharma**
Parse sizes attribute implementation.
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #11416  (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17808)

### CODE REVIEW

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
     }
 }
 
+pub fn parse_a_sizes_attribute(input: DOMString, width: Option<u32>) -> Result<Vec<Size>, String> {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Since we only return Ok from this method (and not Err), there is no reason to return a Result value.

File: components/script/dom/htmlimageelement.rs [hunk 2]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+let mut sizes = Vec::<Size>::new();
+let unparsed_sizes = input.deref().split(',').collect::<Vec<_>>();
+
+    for unparsed_size in &unparsed_sizes {
+        let whitespace = unparsed_size.chars().rev().take_while(|c| char::is_whitespace(*c)).count();
+        let trimmed: String = unparsed_size.chars().take(unparsed_size.chars().count() - whitespace).collect();
+
+        if trimmed.is_empty() {
+            // return Err("reason".to_owned());
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Please remove code instead of commenting it out.

File: components/script/dom/htmlimageelement.rs [hunk 3]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+                    },
+                    Err(_) => {
+                        // return Err("reason".to_owned());
+                        continue;
+                    },
+                }
+            },
+        }
+    }
+    if sizes.len() == 0 {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: nit: use `sizes.is_empty()` instead.

File: tests/unit/script/htmlimageelement.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+
+use script::test::sizes::parse_a_sizes_attribute;
+use script::test::DOMString;
+
+#[test]
+fn empty_vector() {
+    assert!(parse_a_sizes_attribute(DOMString::new(), None).is_ok());
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: We will want to assert that the result is an empty vector (as the test name implies ;))

File: tests/unit/script/htmlimageelement.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+} 
+
+#[test]
+fn test_whitespace() {
+    assert!(parse_a_sizes_attribute(DOMString::from("     (min-width: 500px)"),
+            None).is_ok());
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: Let's think about what this test is checking. `(min-width: 500px)` by itself is not a valid size; it is a media query, and the sizes algorithm then tries to parse a length that follows it. Since there is no length present, this string would be considered a parse error. That's fine, but we should also verify in a separate test that a valid size with extra whitespace parses correctly and gives us a real parsed size.

File: components/script/dom/htmlimageelement.rs [hunk 6]
```diff
@@ -793,6 +873,12 @@ impl HTMLImageElementMethods for HTMLImageElement {
     // https://html.spec.whatwg.org/multipage/#dom-img-src
     make_setter!(SetSrc, "src");
 
+    // https://html.spec.whatwg.org/multipage/images.html#parse-a-sizes-attribute
+    make_url_getter!(Sizes, "sizes");
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: As previously mentioned, this is not a url value.

File: components/script/dom/htmlimageelement.rs [hunk 7]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+            continue;
+        }
+        let mut input = ParserInput::new(&trimmed);
+        let url = ServoUrl::parse("about:blank").unwrap();
+        let reporter = NullReporter;
+        let context = ParserContext::new_for_cssom(&url,
+                                                        &reporter,
+                                                        None,
+                                                        ParsingMode::empty(),
+                                                        QuirksMode::NoQuirks);
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: nit: these lines should line up with the `&url` on the previous line.

File: components/script/dom/htmlimageelement.rs [hunk 8]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
     }
 }
 
+pub fn parse_a_sizes_attribute(input: DOMString, width: Option<u32>) -> Result<Vec<Size>, String> {
+let mut sizes = Vec::<Size>::new();
+let unparsed_sizes = input.deref().split(',').collect::<Vec<_>>();
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: nit: indent these lines to match the `for` below them.

File: components/script/dom/htmlimageelement.rs [hunk 9]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+                length: len,
+                query: None
+            }),
+            Err(_) => {
+                // return Err("reason".to_owned());
+                let mut media_query_parser = Parser::new(&mut input);
+                let media_query = media_query_parser.try(|i| MediaQuery::parse(&context, i));
+                match media_query {
+                    Ok(query) => {
+                        let length = media_query_parser.try(|i| Length::parse_non_negative(&context, i));
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We can use `Length::parse_non_negative` directly instead of `Parser::try`: `let length = Length::parse_non_negative(&context, &mut media_query_parser);`

File: components/script/dom/htmlimageelement.rs [hunk 10]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+                        if length.is_ok() {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Instead of using is_ok and unwrap, we can write `if let Ok(length) = length {`.

File: components/script/dom/htmlimageelement.rs [hunk 11]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+            Ok(len) => sizes.push(Size {
+                length: len,
+                query: None
+            }),
+            Err(_) => {
+                // return Err("reason".to_owned());
+                let mut media_query_parser = Parser::new(&mut input);
+                let media_query = media_query_parser.try(|i| MediaQuery::parse(&context, i));
+                match media_query {
+                    Ok(query) => {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Since the `Err` branch of this match is empty except for `continue`, we can write `if let Ok(query) = media_query {` and omit it entirely.

File: components/script/dom/htmlimageelement.rs [hunk 12]
```diff
@@ -66,6 +76,11 @@ enum State {
     CompletelyAvailable,
     Broken,
 }
+#[allow(dead_code)]
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: This probably is not considered dead code any more, so we can remove it.

File: components/script/dom/htmlimageelement.rs [hunk 13]
```diff
@@ -3,6 +3,8 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use app_units::{Au, AU_PER_PX};
+use core::ops::Deref;
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: This can be removed.

**Review by jdm — CHANGES_REQUESTED** [review]
Good start!

File: components/script/dom/htmlimageelement.rs (hunk 13 above)
Comment (Rakhisharma) [inline components/script/dom/htmlimageelement.rs]: we are using this at line: 802:` let unparsed_sizes = input.deref().split(',').collect::<Vec<_>>();`

Comment (KiChjang) [inline components/script/dom/htmlimageelement.rs]: An explicit call to deref should not be necessary, if `let unparsed_sizes = input.split(',').collect::<Vec<_>>();` does not work, try `let unparsed_sizes = (*input).split(',').collect::<Vec<_>>();`.

File: tests/unit/script/htmlimageelement.rs (hunk 4 above)
Comment (creativcoder) [inline tests/unit/script/htmlimageelement.rs]: @Rakhisharma  As discussed, the test case depicts the situation where sizes is an empty string (`DOMString::new()`), so here we should also check if the vector contains a single `Size` with the value `100vw`. (i.e., the default value returned from spec algorithm)

File: components/script/dom/htmlimageelement.rs [hunk 14]
```diff
@@ -782,6 +797,71 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+        }
+        let mut input = ParserInput::new(&trimmed);
+        let url = ServoUrl::parse("about:blank").unwrap();
+        let reporter = NullReporter;
+        let context = ParserContext::new_for_cssom(&url,
+                                                        &reporter,
+                                                        None,
+                                                        ParsingMode::empty(),
+                                                        QuirksMode::NoQuirks);
+        let url = ServoUrl::parse("about:blank").unwrap();
```
Comment (creativcoder) [inline components/script/dom/htmlimageelement.rs]: You already have `url` declared in line [813](https://github.com/Rakhisharma/servo/blob/bc95d0a8d75e5fbc943985e1e82e588c8d802d79/components/script/dom/htmlimageelement.rs#L813). So you can remove this.

File: tests/unit/script/lib.rs [hunk 15]
```diff
@@ -12,4 +12,6 @@ extern crate servo_url;
 #[cfg(test)] mod textinput;
 #[cfg(test)] mod headers;
 #[cfg(test)] mod htmlareaelement;
+#[cfg(test)] mod htmlimageelement; 
+
```
Comment (creativcoder) [inline tests/unit/script/lib.rs]: nit: Remove the extra newline.

File: components/script/dom/htmlimageelement.rs (hunk 13 above)
Comment (Rakhisharma) [inline components/script/dom/htmlimageelement.rs]: `let unparsed_sizes = input.split(',').collect::<Vec<_>>();` worked! thanks.

File: components/script/dom/htmlimageelement.rs [hunk 16]
```diff
@@ -782,6 +798,63 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+        }
+        let mut input = ParserInput::new(&trimmed);
+        let url = ServoUrl::parse("about:blank").unwrap();
+        let reporter = NullReporter;
+        let context = ParserContext::new_for_cssom(&url,
+                                                   &reporter,
+                                                   None,
+                                                   ParsingMode::empty(),
+                                                   QuirksMode::NoQuirks);
+        let length = Parser::new(&mut input).try(|i| Length::parse_non_negative(&context, i));
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We should create a separate parser object that we can reuse later.

File: components/script/dom/htmlimageelement.rs [hunk 17]
```diff
@@ -782,6 +798,63 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+        println!("\n{:?}", length);
+        match length {
+            Ok(len) => sizes.push(Size {
+                length: len,
+                query: None
+            }),
+            Err(_) => {
+                let mut media_query_parser = Parser::new(&mut input);
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We can reuse the earlier parser instead.

File: components/script/dom/htmlimageelement.rs [hunk 18]
```diff
@@ -782,6 +798,63 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+                let media_query = media_query_parser.try(|i| MediaQuery::parse(&context, i));
+                if let Ok(query) = media_query {
+                        let length = Length::parse_non_negative(&context, &mut media_query_parser);
+                        if length.is_ok() {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: We can use `if let Ok(length) = length {` here and avoid calling unwrap below.

File: components/script/dom/htmlimageelement.rs [hunk 19]
```diff
@@ -782,6 +798,63 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
     }
 }
 
+pub fn parse_a_sizes_attribute(input: DOMString, width: Option<u32>) -> Vec<Size> {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: Add a comment like `// https://html.spec.whatwg.org/multipage/images.html#parse-a-sizes-attribute` so it's clear which algorithm this is based on.

File: components/script/dom/htmlimageelement.rs [hunk 20]
```diff
@@ -793,6 +866,12 @@ impl HTMLImageElementMethods for HTMLImageElement {
     // https://html.spec.whatwg.org/multipage/#dom-img-src
     make_setter!(SetSrc, "src");
 
+    // https://html.spec.whatwg.org/multipage/#parse-a-sizes-attribute
+    make_getter!(Sizes, "sizes");
+    // https://html.spec.whatwg.org/multipage/#parse-a-sizes-attribute
+    make_setter!(SetSizes, "sizes");
+
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs]: nit: remove this extra newline.

File: components/script/test.rs [hunk 21]
```diff
@@ -15,6 +15,33 @@ pub mod area {
...
+pub mod sizes {
+    use app_units::Au;
+    use style::media_queries::{MediaQuery, MediaQueryType};
+    use style::media_queries::Expression;
+    use style::servo::media_queries::{ExpressionKind, Range};
+    use style::values::specified::{Length, NoCalcLength, AbsoluteLength};
+
+    pub use dom::htmlimageelement::{parse_a_sizes_attribute, Size};
+
+    pub fn test_media_query(len: f32) -> MediaQuery {
```
Comment (jdm) [inline components/script/test.rs]: These functions should live in tests/unit/script/htmlimageelement.rs instead of here.

File: tests/unit/script/htmlimageelement.rs [hunk 22]
```diff
@@ -0,0 +1,60 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use script::test::DOMString;
+use script::test::sizes::{parse_a_sizes_attribute, Size, test_media_query, test_length};
+
+#[test]
+fn no_default_provided() {
+    assert!(parse_a_sizes_attribute(DOMString::new(), None).len() == 1);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: Instead of asserting the length, we should be asserting the actual value that is returned in all of these tests.

File: tests/unit/script/htmlimageelement.rs [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+#[test]
+fn default_provided() {
+    assert!(parse_a_sizes_attribute(DOMString::new(), Some(2)).len() == 1);
+}
+
+#[test]
+fn no_size() {
+    assert!(parse_a_sizes_attribute(DOMString::new(), None).len() == 1);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: This is the same as no_default_provided.

File: tests/unit/script/htmlimageelement.rs [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+#[test]
+fn no_extra_whitespace() {
+    let mut a = vec![];
+    let media_query = test_media_query(200f32);
+    let length = test_length(545f32);
+    let size = Size { query: Some(media_query), length: length };
+    a.push(size);
+    assert_eq!(parse_a_sizes_attribute(DOMString::from(" (max-width: 200px) 545px"), None), a);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: It's confusing how this is called `no_extra_whitespace` but the test includes a leading space.

File: tests/unit/script/htmlimageelement.rs [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+    assert_eq!(parse_a_sizes_attribute(DOMString::from("(max-width: 200px) 545px"), None), a);
+}
+
+#[test]
+fn more_then_one_value() {
+    let result = parse_a_sizes_attribute(DOMString::from("(min-width: 900px) 1000px,
+            (max-width: 900px) and (min-width: 400px) 50em,
+            100vw           "),
+            None);
+    assert_eq!(result.len(), 3);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: We should definitely be checking the actual values that are returned here.

File: tests/unit/script/htmlimageelement.rs [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+    assert_eq!(parse_a_sizes_attribute(DOMString::from(" (max-width: 200px) 545px"), None), a);
+}
+
+fn extra_whitespace() {
+    let mut a = vec![];
+    let media_query = test_media_query(200f32);
+    let length = test_length(545f32);
+    let size = Size { query: Some(media_query), length: length };
+    a.push(size);
+    assert_eq!(parse_a_sizes_attribute(DOMString::from("(max-width: 320px) 280px,     (max-width: 480px) 440px, 800px"), None), a);
```
Comment (jdm) [inline tests/unit/script/htmlimageelement.rs]: Why does this test pass? We are parsing a string that contains three different sizes, but the expected results only contain one size that does not match any of them.

**Review by jdm — CHANGES_REQUESTED** [review]
This is a good start! The tests can use a bit more refining, but I think we're close to having something that's ready to be merged!

**Comment by KiChjang** [comment]
Sorry, but which issue does this PR solve?

**Comment by creativcoder** [comment]
This is part of  [#11416](https://github.com/servo/servo/issues/11416)

File: components/script/dom/htmlimageelement.rs (hunk 16 above)
Comment (Rakhisharma) [reply components/script/dom/htmlimageelement.rs]: How to create a separate parser object here? I am not clear if we can create a Parser object without any input string and reuse it later.

File: tests/unit/script/htmlimageelement.rs (hunk 24 above)
Comment (Rakhisharma) [reply tests/unit/script/htmlimageelement.rs]: Here what you mean with the extra whitespace, "does the middle spaces also count as extra whitespace or only the leading and trailing spaces counts as extra whitespace ?"

File: components/script/dom/htmlimageelement.rs (hunk 16 above)
Comment (jdm) [reply components/script/dom/htmlimageelement.rs]:
```rust
let mut parser = Parser::new(&mut input);
/* ... */
let length = parser.try(..);
/* .. */
let something_else = parser.foo(..).
```

File: tests/unit/script/htmlimageelement.rs (hunk 24 above)
Comment (jdm) [reply tests/unit/script/htmlimageelement.rs]: Extra whitespace generally means any additional whitespace beyond what is necessary to separate two tokens, regardless of where the whitespace exists in the string.

File: components/script/test.rs (hunk 21 above)
Comment (Rakhisharma) [reply components/script/test.rs]: Adding this function giving lot of errors https://gist.github.com/Rakhisharma/a28c6a01f3e3a0b6b49fe90948e9fabe  Do you still want me to refactor this function in tests/unit/script/htmlimageelement.rs ?

Comment (jdm) [reply components/script/test.rs]: Right, you need to change the use statements to reference the new location instead of `script::test`.

Comment (jdm) [reply components/script/test.rs]: And when I said these functions, I meant test_media_query and test_length in particular.

File: components/script/dom/htmlimageelement.rs [hunk 27]
```diff
@@ -66,6 +74,12 @@ enum State {
     CompletelyAvailable,
     Broken,
 }
+#[derive(PartialEq)]
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Nit: newline before types please

File: components/script/dom/htmlimageelement.rs [hunk 28]
```diff
@@ -780,6 +794,61 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
     }
 }
 
+//https://html.spec.whatwg.org/multipage/#parse-a-sizes-attribute
+pub fn parse_a_sizes_attribute(input: DOMString, width: Option<u32>) -> Vec<Size> {
+    let mut sizes = Vec::<Size>::new();
+    let unparsed_sizes = input.split(',').collect::<Vec<_>>();
+    for unparsed_size in &unparsed_sizes {
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Since `unparsed_sizes` is used only in the loop, let's iterate on `input.split` directly (which avoids an iteration and vector allocation).

File: components/script/dom/htmlimageelement.rs [hunk 29]
```diff
@@ -780,6 +794,61 @@ impl LayoutHTMLImageElementHelpers for LayoutJS<HTMLImageElement> {
...
+        match length {
+            Ok(len) => sizes.push(Size {
+                length: len,
+                query: None
+            }),
+            Err(_) => {
+                let mut media_query_parser = parser;
+                let media_query = media_query_parser.try(|i| MediaQuery::parse(&context, i));
+                if let Ok(query) = media_query {
+                        let length = Length::parse_non_negative(&context, &mut media_query_parser);
```
Comment (wafflespeanut) [inline components/script/dom/htmlimageelement.rs]: Nit: Improper indentation (should be 4 spaces instead of 8).

File: tests/unit/script/htmlimageelement.rs [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+#[test]
+fn no_default_provided() {
+    let mut a = vec![];
+    let length = test_length_for_no_default_provided(100f32);
+    let size = Size { query: None, length: length };
+    a.push(size);
+    assert_eq!(parse_a_sizes_attribute(DOMString::new(), None), a);
+    println!("{:?}", parse_a_sizes_attribute(DOMString::new(), None));
```
Comment (wafflespeanut) [inline tests/unit/script/htmlimageelement.rs]: Nit: Please remove `println!` usage.

Note for the future: `println!` is not useful in the testsuite because the stdout is captured by cargo. Instead, we should `panic!` so that cargo prints the captured stdout (which also has the panic message).

File: tests/unit/script/htmlimageelement.rs [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+#[test]
+fn default_provided() {
+    let mut a = vec![];
+    let length = test_length_for_default_provided(2f32);
+    let size = Size { query: None, length: length };
+    a.push(size);
+    assert_eq!(parse_a_sizes_attribute(DOMString::new(), Some(2)), a);
+    println!("{:?}", parse_a_sizes_attribute(DOMString::new(), Some(2)));
```
Comment (wafflespeanut) [inline tests/unit/script/htmlimageelement.rs]: Nit: Ditto on `println!`

File: tests/unit/script/htmlimageelement.rs [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+    let length = Length::NoCalc(NoCalcLength::Absolute(AbsoluteLength::Px(len)));
+    return length;
+}
+
+#[test]
+fn default_provided() {
+    let mut a = vec![];
+    let length = test_length_for_default_provided(2f32);
+    let size = Size { query: None, length: length };
+    a.push(size);
```
Comment (wafflespeanut) [inline tests/unit/script/htmlimageelement.rs]: Nit: If it's just a single size, then let's have `let a = vec![size];` (here and everywhere below)

**Review by ghost — COMMENTED** [review]
Nice work! This mostly looks good to me. Just a few nits, and we'll be good to go.

File: components/script/dom/htmlimageelement.rs (hunk 28 above)
Comment (Rakhisharma) [reply components/script/dom/htmlimageelement.rs]: Not sure what you want me to do here. Would you please elaborate little more. Thanks :)

Comment (wafflespeanut) [reply components/script/dom/htmlimageelement.rs]: I meant,

``` rust
for unparsed_size in input.split(',') {
```

This way, we perform lazy iteration.

**Comment by jdm** [comment]
I went ahead and squashed these commits and opened a new PR in #18714.
